[JSC] Implement RegExp.prototype[Symbol.replace] in C++

Constellation opened the pull request:
#### 55b14a056ad3a98894f23c81dfc09057c836ca65
<pre>
[JSC] Implement RegExp.prototype[Symbol.replace] in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=304752">https://bugs.webkit.org/show_bug.cgi?id=304752</a>
<a href="https://rdar.apple.com/167280055">rdar://167280055</a>

Reviewed by Sosuke Suzuki.

This patch implements RegExp.prototype[Symbol.replace] in C++ instead of
JS. This is not super important function since String.prototype.replace
and its fast path is mostly called, and this function is rarely called.
Let&apos;s just have it in C++ so that we can simplify our builtin
implementations more.
We generate bunch of tests for RegExp.prototype[Symbol.replace] to cover
the cases.

Tests: JSTests/stress/regexp-prototype-symbol-replace-basic.js
       JSTests/stress/regexp-prototype-symbol-replace-capture-groups.js
       JSTests/stress/regexp-prototype-symbol-replace-edge-cases.js
       JSTests/stress/regexp-prototype-symbol-replace-functional.js
       JSTests/stress/regexp-prototype-symbol-replace-named-groups.js
       JSTests/stress/regexp-prototype-symbol-replace-slow-path.js
       JSTests/stress/regexp-prototype-symbol-replace-unicode.js

* JSTests/stress/regexp-prototype-symbol-replace-basic.js: Added.
(shouldBe):
* JSTests/stress/regexp-prototype-symbol-replace-capture-groups.js: Added.
(shouldBe):
* JSTests/stress/regexp-prototype-symbol-replace-edge-cases.js: Added.
(shouldBe):
(shouldThrow):
(toString):
(let.fakeRegExp.exec):
* JSTests/stress/regexp-prototype-symbol-replace-functional.js: Added.
(shouldBe):
(return.toString):
* JSTests/stress/regexp-prototype-symbol-replace-named-groups.js: Added.
(shouldBe):
* JSTests/stress/regexp-prototype-symbol-replace-slow-path.js: Added.
(shouldBe):
(reWithCustomExec.exec):
(customResult.exec):
(customLength.exec):
(modifyingExec.exec):
(nullExec.exec):
(let.plainObject.exec):
(throwingExec.exec):
(i.i.2.0.re.exec):
(funcWithCustomExec.exec):
(slowPathRe.exec):
(namedCapturesExec.exec):
* JSTests/stress/regexp-prototype-symbol-replace-unicode.js: Added.
(shouldBe):
(catch):
* Source/JavaScriptCore/builtins/RegExpPrototype.js:
(linkTimeConstant.getSubstitution): Deleted.
(overriddenName.string_appeared_here.replace): Deleted.
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::addStringReplacePrimordialChecks):
* Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
(JSC::advanceStringUnicode):
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::RegExpPrototype::finishCreation):
(JSC::advanceStringIndex):
(JSC::getSubstitution):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/304983@main">https://commits.webkit.org/304983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/902757eecf6a3578046cf0571c25ec5a7756c2f7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9454 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48381 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144837 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90067 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e3447265-b48c-450e-b261-a83e138cff29) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10157 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9581 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104838 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a41b49af-fd2c-4b7b-a758-5b084accfe21) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140039 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7467 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122849 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85673 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/425d4fe2-0db2-4d05-9e86-081126af0761) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7104 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4810 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5427 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/129055 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116452 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41017 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147593 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135581 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9137 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41582 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113192 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9155 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7679 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113522 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7026 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119112 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63477 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21123 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9185 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37166 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168361 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8909 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72751 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43944 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9126 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8978 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on JSTests/stress/regexp-prototype-symbol-replace-basic.js:
```diff
@@ -0,0 +1,62 @@
+shouldBe(/\d+/g[Symbol.replace]("abc123def456", "XXX"), "abcXXXdefXXX");
+
+// Test 14: Repeated application (JIT warmup)
+for (let i = 0; i < 1000; i++) {
```
```suggestion
for (let i = 0; i < testLoopCount; i++) {
```

sosukesuzuki commented on JSTests/stress/regexp-prototype-symbol-replace-edge-cases.js:
```diff
@@ -0,0 +1,134 @@
+shouldBe(/a/gy[Symbol.replace]("aaab", "X"), "XXXb");
+
+// Test 21: JIT warmup edge cases
+for (let i = 0; i < 1000; i++) {
```
ditto

sosukesuzuki commented on JSTests/stress/regexp-prototype-symbol-replace-named-groups.js:
```diff
@@ -0,0 +1,54 @@
+shouldBe(/(?<word>\w+)/[Symbol.replace]("hello world", "$<word>$&$<word>"), "hellohellohello world");
+
+// Test 12: JIT warmup with named captures
+for (let i = 0; i < 1000; i++) {
```
ditto

sosukesuzuki commented on JSTests/stress/regexp-prototype-symbol-replace-unicode.js:
```diff
@@ -0,0 +1,70 @@
+}
+
+// Test 14: JIT warmup with unicode
+for (let i = 0; i < 1000; i++) {
```
ditto

sosukesuzuki reviewed (approved):
r=me

sosukesuzuki reviewed (changes requested):
test262-runner failed
```
! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/y-init-lastindex.js (strict mode) (Exit code: 3)
Exception: Test262Error: Expected SameValue(«"xaa"», «"axa"») to be true

! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/y-fail-lastindex.js (strict mode) (Exit code: 3)
Exception: Test262Error: Expected SameValue(«1», «0») to be true

! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/y-init-lastindex.js (default) (Exit code: 3)
Exception: Test262Error: Expected SameValue(«"xaa"», «"axa"») to be true

! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/y-fail-lastindex.js (default) (Exit code: 3)
Exception: Test262Error: Expected SameValue(«1», «0») to be true

! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/y-fail-lastindex-no-write.js (strict mode) (Exit code: 3)
Exception: Test262Error: Expected a TypeError to be thrown but no exception was thrown at all

! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/y-fail-lastindex-no-write.js (default) (Exit code: 3)
Exception: Test262Error: Expected a TypeError to be thrown but no exception was thrown at all

[755/1! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/coerce-lastindex.js (strict mode) (Exit code: 3)
Exception: Test262Error: Expected SameValue(«0», «9007199254740992») to be true

! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/y-set-lastindex.js (strict mode) (Exit code: 3)
Exception: Test262Error: Expected SameValue(«0», «3») to be true

! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/coerce-lastindex.js (default) (Exit code: 3)
Exception: Test262Error: Expected SameValue(«0», «9007199254740992») to be true

! NEW FAIL test/built-ins/RegExp/prototype/Symbol.replace/y-set-lastindex.js (default) (Exit code: 3)
Exception: Test262Error: Expected SameValue(«0», «3») to be true
```

sosukesuzuki reviewed (approved):
r=me

Constellation committed:
[JSC] Implement RegExp.prototype[Symbol.replace] in C++
https://bugs.webkit.org/show_bug.cgi?id=304752
rdar://167280055

Reviewed by Sosuke Suzuki.

This patch implements RegExp.prototype[Symbol.replace] in C++ instead of
JS. This is not super important function since String.prototype.replace
and its fast path is mostly called, and this function is rarely called.
Let's just have it in C++ so that we can simplify our builtin
implementations more.
We generate bunch of tests for RegExp.prototype[Symbol.replace] to cover
the cases.

Tests: JSTests/stress/regexp-prototype-symbol-replace-basic.js
       JSTests/stress/regexp-prototype-symbol-replace-capture-groups.js
       JSTests/stress/regexp-prototype-symbol-replace-edge-cases.js
       JSTests/stress/regexp-prototype-symbol-replace-functional.js
       JSTests/stress/regexp-prototype-symbol-replace-named-groups.js
       JSTests/stress/regexp-prototype-symbol-replace-slow-path.js
       JSTests/stress/regexp-prototype-symbol-replace-unicode.js

* JSTests/stress/regexp-prototype-symbol-replace-basic.js: Added.
(shouldBe):
* JSTests/stress/regexp-prototype-symbol-replace-capture-groups.js: Added.
(shouldBe):
* JSTests/stress/regexp-prototype-symbol-replace-edge-cases.js: Added.
(shouldBe):
(shouldThrow):
(toString):
(let.fakeRegExp.exec):
* JSTests/stress/regexp-prototype-symbol-replace-functional.js: Added.
(shouldBe):
(return.toString):
* JSTests/stress/regexp-prototype-symbol-replace-named-groups.js: Added.
(shouldBe):
* JSTests/stress/regexp-prototype-symbol-replace-slow-path.js: Added.
(shouldBe):
(reWithCustomExec.exec):
(customResult.exec):
(customLength.exec):
(modifyingExec.exec):
(nullExec.exec):
(let.plainObject.exec):
(throwingExec.exec):
(i.i.2.0.re.exec):
(funcWithCustomExec.exec):
(slowPathRe.exec):
(namedCapturesExec.exec):
* JSTests/stress/regexp-prototype-symbol-replace-unicode.js: Added.
(shouldBe):
(catch):
* Source/JavaScriptCore/builtins/RegExpPrototype.js:
(linkTimeConstant.getSubstitution): Deleted.
(overriddenName.string_appeared_here.replace): Deleted.
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::addStringReplacePrimordialChecks):
* Source/JavaScriptCore/runtime/RegExpObjectInlines.h:
(JSC::advanceStringUnicode):
* Source/JavaScriptCore/runtime/RegExpPrototype.cpp:
(JSC::RegExpPrototype::finishCreation):
(JSC::advanceStringIndex):
(JSC::getSubstitution):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/304983@main
